### PR TITLE
Insurance Product Core (IPC) API tests [CM-726, CM-727, CM-728, CM-729, CM-730, CM-747]

### PR DESCRIPTION
**Pull request by KellenKincaid**
This PR contains tests for the Insurance Product Core (IPC, formerly known as IPIM) API endpoints.
The tests verify 2 things:

1. The API endpoints work as expected
2. The data retrieved by the API endpoints matches the data in [config/us_states.yml](https://github.com/svinstech/pa_shim/blob/main/config/us_states.yml) within the pa_shim repo.

These tests were requested by the Carrier Management team.

### CODE REVIEW

**Comment by KellenKincaid** [comment]
Currently, the check is failing because there is a [bug](https://vouchinc.atlassian.net/browse/CM-741).

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+          command: yarn run test:jurisdictions
+      - run:
+          name: Run IPC Jurisdiction Version tests
+          command: yarn run test:jurisdictionVersions
+      - run:
+          name: Run IPC Product Lines tests
+          command: yarn run test:productLines
+      - slack-notifier/build-status-notification: 
+          header: "Insurance Product Core Tests"
+          channel-webhook-environment-variables: WEBHOOK_KELLEN_KINCAID
```
Comment (c0nniewang) [inline .circleci/config.yml:50]: what slack channel does this alert to?

File: README.md [hunk 2]
```diff
@@ -1,2 +1,33 @@
...
+`yarn run test:health`
+`yarn run test:jurisdictions`
+`yarn run test:jurisdictionVersions`
+`yarn run test:productLines`
+
+### RUNNING TESTS IN CIRCLECI
+- In the CircleCI config, ensure that the job running the tests has: `resource_class: svinstech/dev`
+    - This will fill the same role as being on the VPN when it comes to endpoint access.
+- Ensure that you set the target environment to staging, as this is the only environment that `svinstech/dev` has access to.
+- The tests are scheduled to automatically run in `main` every Wednesday at midnight EST (4am UTC)
```
Comment (c0nniewang) [inline README.md:33]: is this convention for our other automated tests? would it make sense to run these tests on every uat and/or prod deploy or is that unnecessary do you think?

File: index.ts [hunk 3]
```diff
@@ -1,10 +1,13 @@
...
-import { readFileSync, writeFileSync } from 'fs'
-import axios from 'axios'
-import { AxiosResponse } from 'axios'
-import { PaShimUsStatesYaml } from './src/interfaces/interfacesAndTypes'
+import { Endpoint, Url } from './src/constants'
+import { GetPaShimUsStateData } from './src/dataCollection'
+import { Jurisdiction, JurisdictionVersion, ProductLine, UsStateMapping } from './src/interfaces/interfacesAndTypes'
 
 async function main() :Promise<void> {
+    ///////////////////////////////////////////////////////////////////////////////
```
Comment (c0nniewang) [inline index.ts:6]: do we want to keep this and the deleted lines below ?

File: src/dataCollection.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
+import axios from 'axios'
+import { AxiosResponse } from 'axios'
+import { load } from 'js-yaml'
+import { readFileSync, writeFileSync, existsSync, mkdirSync } from 'fs'
+import { Jurisdiction, JurisdictionVersion, PaShimUsStatesYaml, ProductLine, TestData, UsStateMapping, DataAndStatus } from './interfaces/interfacesAndTypes'
+import { Url, Endpoint } from './constants'
+
+
```
Comment (c0nniewang) [inline src/dataCollection.ts:8]: np: extra line here

File: src/dataCollection.ts [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+/*
+    Returns true if the given response code is a 2xx code. False otherwse.
+*/
+export function IsGoodResponse(_responseCode:number) :boolean {
+    const output:boolean = _responseCode >= 200 && _responseCode < 300
+
+    if (!output) {
+        //testing
```
Comment (c0nniewang) [inline src/dataCollection.ts:16]: should this be removed?

File: src/dataCollection.ts [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+    These 
+*/
+export async function GetPaShimUsStateData(_fileNameWithoutExtension?:string) :Promise<DataAndStatus<UsStateMapping>> {
+    _fileNameWithoutExtension ??= "paShimUsStates"
+
+    const GITHUB_PERSONAL_ACCESS_TOKEN = process.env.GITHUB_TOKEN,
+          collectedDataDirectoryName = "collected_data",
+          paShimUsStatesYamlFileName = `${collectedDataDirectoryName}/${_fileNameWithoutExtension}.yml`,
+          paShimUsStatesJsonFileName = `${collectedDataDirectoryName}/${_fileNameWithoutExtension}.json`,
+          yamlUrl = "https://api.github.com/repos/svinstech/pa_shim/git/blobs/3a81ca37fc0df296ea76e60807b5d6e4a9468b73" // Points to the config/us_states.yml in pa_shim.
```
Comment (c0nniewang) [inline src/dataCollection.ts:39]: this is going to break once we introduce new jurisdictions/jurisdiction versions in IPC, corresponding changes in pa_shim will be stored in active record not the yml file

File: src/dataCollection.ts [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+export async function HitEndpoint(_baseUrl:Url, _endpoint:Endpoint|string) :Promise<DataAndStatus<any>> {
+    const url :string = `${_baseUrl}/${_endpoint}`;
+    const data :DataAndStatus<any> = await GetResponseData(url);
+
+    return data;
+}
+
+
+
+
```
Comment (c0nniewang) [inline src/dataCollection.ts:130]: np: extra lines here

File: src/dataCommunication.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
+
+
```
Comment (c0nniewang) [inline src/dataCommunication.ts:2]: np: extra lines here

File: src/dataManipulation.ts [hunk 9]
```diff
@@ -0,0 +1,60 @@
+import { Jurisdiction, JurisdictionVersion, ProductLine, UsState, UsStateMapping } from "./interfaces/interfacesAndTypes";
+
+
```
Comment (c0nniewang) [inline src/dataManipulation.ts:3]: np: extra line

File: src/dataCommunication.ts [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+/*
+    This is meant to receive an array that contains the difference between 2 arrays.
+    It logs that array after explaining the problem that it represents.
+*/
+export function LogArrayDifference(_differenceArray:string[]) :void {
+    if (_differenceArray.length > 0) {
+        const theseOrThis = (_differenceArray.length > 1) ? "These are" : "This is"
+        console.log(`DISCREPANCY: ${theseOrThis} not shared between pa_shim & IPC: `)
```
Comment (c0nniewang) [inline src/dataCommunication.ts:10]: not totally sure how this works in the circleci runner but will we be able to see these console.log's in envs other than our local env?

File: src/dataManipulation.ts [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+/*
+    Takes the pa_shim version data and returns an array of its unique states.
+*/
+export function GetUniqueArrayOfPaShimStates(_paShimStatesData:UsStateMapping) :UsState[] {
+    return Object.keys(_paShimStatesData as object) as UsState[];
+}
+
+
+
+
```
Comment (c0nniewang) [inline src/dataManipulation.ts:60]: np: extra lines

File: src/test/health.spec.ts [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+import { Endpoint, ENVIRONMENT } from "../constants";
+import { HitEndpoint, IsGoodResponse } from "../dataCollection";
+import { DataAndStatus } from "../interfaces/interfacesAndTypes";
+
+describe("~~~ HEALTH ~~~", () => {
+    it("Ensure that the /health endpoint works and is ok.", async () => {
+        const healthResponse:DataAndStatus<string> = await HitEndpoint(ENVIRONMENT, Endpoint.Health) as DataAndStatus<string>
+
+        //testing
+        console.log(`health: ${healthResponse}`)
```
Comment (c0nniewang) [inline src/test/health.spec.ts:12]: should these be removed?

File: src/test/jurisdictionVersions.spec.ts [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+    
+            //testing
+            console.log("paShimStates")
+            console.log(paShimStates)
+
+            const ipcStates:UsState[] = GetUniqueArrayOfIpcStates(JURISDICTION_VERSION_RESPONSE.data);
+    
+            //testing
+            console.log("ipcStates")
+            console.log(ipcStates)
```
Comment (c0nniewang) [inline src/test/jurisdictionVersions.spec.ts:40]: same with these console logs and above, should they be removed?

File: src/test/jurisdictionVersions.spec.ts [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+                const specificJurisdictionVersion: DataAndStatus<JurisdictionVersion> = response as DataAndStatus<JurisdictionVersion>
+                expect(IsGoodResponse(specificJurisdictionVersion.status)).to.be.false
+
+                expect(JURISDICTION_VERSION_RESPONSE.data).to.not.include(specificJurisdictionVersion.data)
+            })
+        })
+    })
+})
+
+
```
Comment (c0nniewang) [inline src/test/jurisdictionVersions.spec.ts:194]: np: extra line

**Review by c0nniewang — COMMENTED** [review]
overall this lgtm, some of the tests i anticipate breaking once we start adding new JVs or jurisdictions but i think that's okay for now. also it might be nice to run a linter on these files, would help out with some formatting inconsistencies. also there are a lot of console logs in the tests, not sure if you intentionally left those in or not but might be nice to clean that up.

File: .circleci/config.yml (hunk 1 above)
Comment (KellenKincaid) [reply .circleci/config.yml:50]: It DMs me specifically. 
However we can add other recipients whenever necessary. Currently the message just says if the tests all passed or all failed. In the future i plan to make the message a lot more descriptive.

File: README.md (hunk 2 above)
Comment (KellenKincaid) [reply README.md:33]: This is not the convention - I just picked a random day, but we can change this later.

File: src/dataCommunication.ts (hunk 10 above)
Comment (KellenKincaid) [reply src/dataCommunication.ts:10]: In CircleCI, the output of these console.log's will be in same place as the test results.
For example, if you go [here](https://app.circleci.com/pipelines/github/svinstech/ipc_api_testing/35/workflows/1bdd5190-7789-4904-b5da-910dc173e577/jobs/36) and open _Run IPC Jurisdiction Version tests_, you'll see some console.log's in action.

File: src/test/jurisdictionVersions.spec.ts (hunk 14 above)
Comment (KellenKincaid) [reply src/test/jurisdictionVersions.spec.ts]: Removed extra newlines

File: src/test/jurisdictionVersions.spec.ts (hunk 13 above)
Comment (KellenKincaid) [reply src/test/jurisdictionVersions.spec.ts:40]: I'd like to keep these here for peace of mind as well.

File: src/test/health.spec.ts (hunk 12 above)
Comment (KellenKincaid) [reply src/test/health.spec.ts:12]: I'd like to keep this here for peace of mind.

File: src/dataManipulation.ts (hunk 11 above)
Comment (KellenKincaid) [reply src/dataManipulation.ts]: Removed extra newlines

File: src/dataCollection.ts (hunk 5 above)
Comment (KellenKincaid) [reply src/dataCollection.ts]: Removed log.

File: src/dataCollection.ts (hunk 4 above)
Comment (KellenKincaid) [reply src/dataCollection.ts]: Removed extra newline

File: index.ts (hunk 3 above)
Comment (KellenKincaid) [reply index.ts]: File deleted

**Review by c0nniewang — APPROVED** [review]

File: src/dataCollection.ts (hunk 6 above)
Comment (KellenKincaid) [reply src/dataCollection.ts:31]: Thanks, I'll reach out to Ben Watts about how to access Active record.

File: src/dataCollection.ts (hunk 7 above)
Comment (KellenKincaid) [reply src/dataCollection.ts:122]: Removed these empty lines.

File: src/dataCommunication.ts (hunk 8 above)
Comment (KellenKincaid) [reply src/dataCommunication.ts]: Removed these empty lines.

File: src/dataManipulation.ts (hunk 9 above)
Comment (KellenKincaid) [reply src/dataManipulation.ts]: Removed an empty line.

**Comment by KellenKincaid** [comment]
> overall this lgtm, some of the tests i anticipate breaking once we start adding new JVs or jurisdictions but i think that's okay for now. also it might be nice to run a linter on these files, would help out with some formatting inconsistencies. also there are a lot of console logs in the tests, not sure if you intentionally left those in or not but might be nice to clean that up.

Thanks for the review. I will reach out to Ben Watts about how to compensate for the data being migrated to Active Record. 
I will add a linter on a separate PR. I have created a card for this work [here](https://vouchinc.atlassian.net/browse/CM-848).
I've also removed a handful of console logs, but kept some as I believe that it will be important to be able to easily double check what data is being compared in some of the tests.

**Pull request merged**